File: LeetCode/637. Average of Levels in Binary Tree.js
// https://leetcode.com/problems/average-of-levels-in-binary-tree/description

// Более очевидное решение
// const averageOfLevels = (root) => {
//   if (!root) return [];

//   const result = [];

//   let stack = [root];

//   while (stack.length) {
//     const temp = [];

//     const count = stack.length;

//     let currentSum = 0;

//     while (stack.length) {
//       const { val, left, right } = stack.pop();
//       currentSum += val;

//       left && temp.push(left);
//       right && temp.push(right);
//     }

//     stack = temp;

//     result.push(currentSum / count);
//   }

//   return result;
// };

const averageOfLevels = (root) => {
  if (!root) return [];

  const result = [];
  const stack = [root];

  let index = 0;

  let count = 1;

  while (index < stack.length) {
    const currentLevelLength = count;
    let currentLevelEnd = index + count;

    let currentSum = 0;

    count = 0;

    while (index < currentLevelEnd) {
      const { val, left, right } = stack[index];
      currentSum += val;

      if (left) {
        stack.push(left);
        count++;
      }
      if (right) {
        stack.push(right);
        count++;
      }

      index++;
    }

    result.push(currentSum / currentLevelLength);
  }

  return result;
};

File: LeetCode/637. Average of Levels in Binary Tree.py
# https://leetcode.com/problems/average-of-levels-in-binary-tree/description


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def averageOfLevels(self, root: TreeNode | None) -> list[float]:
        if not root:
            return []

        result: list[float] = []

        stack: list[TreeNode] = [root]

        while len(stack) != 0:

            current_row = stack
            stack = []

            current_sum: float = 0
            current_length = len(current_row)

            while len(current_row) != 0:
                current = current_row.pop()

                if current.left:
                    stack.append(current.left)

                if current.right:
                    stack.append(current.right)

                current_sum += current.val

            result.append(current_sum / current_length)

        return result
